src/caches/beliefpropagationcache.jl

```diff
@@ -0,0 +1,221 @@
+  return BeliefPropagationCache(ptn; kwargs...)
+end
+
+partitioned_itensornetwork(bpc::BeliefPropagationCache) = bpc.partitioned_itensornetwork
+message_tensors(bpc::BeliefPropagationCache) = bpc.message_tensors
+function tensornetwork(bpc::BeliefPropagationCache)
+  return unpartitioned_graph(partitioned_itensornetwork(bpc))
+end
+function NamedGraphs.partitioned_graph(bpc::BeliefPropagationCache)
+  return partitioned_graph(partitioned_itensornetwork(bpc))
+end
+
+function initial_message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  ptn = partitioned_itensornetwork(bpc)
+  src_e_itn = subgraph(ptn, src(edge))
+  dst_e_itn = subgraph(ptn, dst(edge))
+  inds_e = commoninds(src_e_itn, dst_e_itn)
+  return mt_constructor(inds_e)
+end
+
+function message_tensor(
+  bpc::BeliefPropagationCache, edge::PartitionEdge; mt_constructor=default_mt_constructor
+)
+  mts = message_tensors(bpc)
+  return haskey(mts, edge) ? mts[edge] : initial_message_tensor(bpc, edge; mt_constructor)
+end
+function message_tensors(
+  bpc::BeliefPropagationCache, edges::Vector{PartitionEdge}; kwargs...
+)
+  return [message_tensor(bpc, edge; kwargs...) for edge in edges]
+end
+
+function copy(bpc::BeliefPropagationCache)
+  return BeliefPropagationCache(
+    copy(partitioned_itensornetwork(bpc)), copy(message_tensors(bpc))
+  )
+end
+
+function default_bp_niters(bpc::BeliefPropagationCache)
+  return default_bp_niters(partitioned_graph(bpc))
+end
+function default_edge_sequence(bpc::BeliefPropagationCache)
+  return default_edge_sequence(partitioned_itensornetwork(bpc))
+end
+
+"""
+Compute message tensor as product of incoming mts and local state
+"""
+function updated_message_tensor(
+  bpc::BeliefPropagationCache,
+  edge::PartitionEdge;
+  contractor=default_contractor,
+  contractor_kwargs=default_contractor_kwargs(),
+  mt_constructor=default_mt_constructor,
+)
+  ptn = partitioned_itensornetwork(bpc)
+  pb_edges = partitionedges(ptn, boundary_edges(ptn, vertices(ptn, src(edge)); dir=:in))
+
+  incoming_messages = [
+    message_tensor(bpc, e_in; mt_constructor) for e_in in setdiff(pb_edges, [reverse(edge)])
+  ]
+  incoming_messages = reduce(vcat, incoming_messages; init=ITensor[])
+
+  contract_list = ITensor[
+    incoming_messages
+    Vector{ITensor}(subgraph(ptn, src(edge)))
+  ]
+
+  return contractor(contract_list; contractor_kwargs...)
+end
+
+"""
+Do a sequential update of the message tensors on `edges`
+"""
+function update(
+  bpc::BeliefPropagationCache,
+  edges::Vector{<:PartitionEdge};
+  compute_norm=false,
+  mt_constructor=default_mt_constructor,
+  kwargs...,
+)
+  bpc_updated = copy(bpc)
+  mts = message_tensors(bpc_updated)
+  c = 0
+  for e in edges
+    mts[e] = updated_message_tensor(bpc_updated, e; mt_constructor, kwargs...)
+    if compute_norm
+      c += message_tensor_diff(message_tensor(bpc, e; mt_constructor), mts[e])
+    end
+  end
+  return bpc_updated, c / (length(edges))
```

Comment:
I think instead of returning the difference like this, we can instead pass it as a keyword argument rapped in a reference, i.e.:
```julia
function update(
  bp_cache::BeliefPropagationCache,
  edges::Vector{<:PartitionEdge};
  (update_diff!)=nothing,
  kwargs...,
)
  # ...
  if !isnothing(update_diff)
    update_diff![] += message_diff(message(bp_cache, edge), messages[edge])
  end
  return bp_cache_updated
end
```
and then it can be called like this:
```julia
diff = Ref(0)
update(bp_cache, edges; (update_diff!)=diff)
```
and then `diff` will store the difference between the updated and original message tensors.